prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -0,0 +1,603 @@
+	return rxr_match_addr(rx_entry->addr, pkt_entry->addr);
+}
+
+static
+int rxr_pkt_rtm_match_trecv(struct dlist_entry *item, const void *arg)
+{
+	struct rxr_pkt_entry *pkt_entry = (struct rxr_pkt_entry *)arg;
+	struct rxr_rx_entry *rx_entry;
+	uint64_t match_tag;
+
+	rx_entry = container_of(item, struct rxr_rx_entry, entry);
+	match_tag = rxr_pkt_rtm_tag(pkt_entry);
+
+	return rxr_match_addr(rx_entry->addr, pkt_entry->addr) &&
+	       rxr_match_tag(rx_entry->cq_entry.tag, rx_entry->ignore,
+			     match_tag);
+}
+
+static
+struct rxr_rx_entry *rxr_pkt_get_msgrtm_rx_entry(struct rxr_ep *ep,
+						 struct rxr_pkt_entry **pkt_entry_ptr)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct dlist_entry *match;
+
+	match = dlist_find_first_match(&ep->rx_list, &rxr_pkt_rtm_match_recv, *pkt_entry_ptr);
+	if (OFI_UNLIKELY(!match)) {
+		/*
+		 * rxr_ep_alloc_unexp_rx_entry_for_msgrtm() might release pkt_entry,
+		 * thus we have to use pkt_entry_ptr here
+		 */
+		rx_entry = rxr_ep_alloc_unexp_rx_entry_for_msgrtm(ep, pkt_entry_ptr);
+		if (OFI_UNLIKELY(!rx_entry))
+			return NULL;
+
+	} else {
+		rx_entry = rxr_pkt_get_rtm_matched_rx_entry(ep, match, *pkt_entry_ptr);
+	}
+
+	assert(rx_entry->total_len > 0);
+	return rx_entry;
+}
+
+static
+struct rxr_rx_entry *rxr_pkt_get_tagrtm_rx_entry(struct rxr_ep *ep,
+						 struct rxr_pkt_entry **pkt_entry_ptr)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct dlist_entry *match;
+
+	match = dlist_find_first_match(&ep->rx_tagged_list, &rxr_pkt_rtm_match_trecv, *pkt_entry_ptr);
+	if (OFI_UNLIKELY(!match)) {
+		/*
+		 * rxr_ep_alloc_unexp_rx_entry_for_tagrtm() might release pkt_entry,
+		 * thus we have to use pkt_entry_ptr here
+		 */
+		rx_entry = rxr_ep_alloc_unexp_rx_entry_for_tagrtm(ep, pkt_entry_ptr);
+		if (OFI_UNLIKELY(!rx_entry))
+			return NULL;
+
+	} else {
+		rx_entry = rxr_pkt_get_rtm_matched_rx_entry(ep, match, *pkt_entry_ptr);
+	}
+
+	return rx_entry;
+}
+
+ssize_t rxr_pkt_proc_matched_rtm(struct rxr_ep *ep,
+				 struct rxr_rx_entry *rx_entry,
+				 struct rxr_pkt_entry *pkt_entry)
+{
+	char *data;
+	size_t data_size, bytes_left;
+	ssize_t ret;
+
+	assert(rx_entry->state == RXR_RX_MATCHED);
+	data = (char *)pkt_entry->pkt + pkt_entry->hdr_size;
+	data_size = pkt_entry->pkt_size - pkt_entry->hdr_size;
+	bytes_left = rxr_pkt_req_copy_data(rx_entry, pkt_entry,
+					   data, data_size);
+	if (!bytes_left) {
+		/*
+		 * rxr_cq_handle_rx_completion() releases pkt_entry, thus
+		 * we do not release it here.
+		 */
+		rxr_cq_handle_rx_completion(ep, pkt_entry, rx_entry);
+		rxr_msg_multi_recv_free_posted_entry(ep, rx_entry);
+		rxr_release_rx_entry(ep, rx_entry);
+		ret = 0;
+	} else {
+		/*
+		 * long message protocol
+		 */
+#if ENABLE_DEBUG
+		dlist_insert_tail(&rx_entry->rx_pending_entry, &ep->rx_pending_list);
+		ep->rx_pending++;
+#endif
+		rx_entry->state = RXR_RX_RECV;
+		rx_entry->tx_id = rxr_get_long_rtm_base_hdr(pkt_entry->pkt)->tx_id;
+		rx_entry->credit_request = rxr_env.tx_min_credits;
```

Comment:
Isn't the credit request logic supposed to be here?

Reply:
Due to a bug in current master branch, the rx_entry->credit_request was always set to rxr_env.tx_min_credits. I am mimicking the behavior of master here.

Reply:
Can we fix it for this release? Or is there a performance impact to doing that?

Reply:
Yes, using the default value has better bandwidth.

Reply:
This needs a comment for that then.

Reply:
Added a comments.